options/options_helper.cc

```diff
@@ -577,6 +595,90 @@ Status GetMutableOptionsFromStrings(
   return Status::OK();
 }
 
+Status ParseOptionsTypeFromMap(
+    const ConfigOptions& config_options,
+    const std::unordered_map<std::string, OptionTypeInfo>& opt_map,
+    void* opt_ptr,
+    const std::unordered_map<std::string, std::string>& options) {
+  std::unordered_map<std::string, std::string> unused_opts;
+  return ParseOptionsTypeFromMap(config_options, opt_map, opt_ptr, options,
+                                 &unused_opts);
+}
+
+Status ParseOptionsTypeFromMap(
+    const ConfigOptions& config_options,
+    const std::unordered_map<std::string, OptionTypeInfo>& opt_map,
+    void* opt_ptr, const std::unordered_map<std::string, std::string>& options,
+    std::unordered_map<std::string, std::string>* unused_opts) {
+  Status s, result, invalid;
+  bool found_one = false;
+  std::unordered_map<std::string, std::string> invalid_opts;
+  // Go through all of the values in the input map and attempt to configure the
+  // property.
+  for (const auto& o : options) {
+    std::string opt_name;
+    const auto opt_info = OptionTypeInfo::Find(o.first, opt_map, &opt_name);
+    if (opt_info == nullptr) {
+      result = Status::NotFound("Could not find option: ", o.first);
+      unused_opts->insert(o);
+    } else {
+      s = opt_info->Parse(config_options, opt_name, o.second, opt_ptr);
+      if (s.ok()) {
+        found_one = true;
+      } else if (s.IsNotFound()) {
+        result = s;
+        unused_opts->insert(o);
+      } else if (s.IsNotSupported()) {
+        // Let not supported appear in unused for the first pass
+        unused_opts->insert(o);
+      } else {
+        invalid_opts.insert(o);
+        invalid = s;
+      }
+    }
+  }
+  // While there are unused properties and we processed at least one,
+  // go through the remaining unused properties and attempt to configure them.
+  while (found_one && !unused_opts->empty()) {
```

Comment:
I'm confused here, if we cannot parse the unused options in the first round, why we are able to parse it in the second round?

Reply:
There are a few ways this can happen.

1.  When we support dynamic loading of objects, it is possible the factories are not loaded during the first pass, meaning we cannot instantiate the instance until we instantiate the registry;

2.  If values are encoded independently (table.id = BlockBasedTable; table.checksum=12345;", it is possible that the table.checksum is processed before table.id, in which case the underlying object may be wrong or not created yet. In this case, the checksum property would not exist until the ID was interpreted.